google/web-stories-wp
# Media: Only transcode 1 resource at a time

## Context



Too many ongoing ffmpeg process can cause the browser to crash due to resource usage being too high.

## Summary



This PR changes the media upload logic so that only 1 video is ever transcoded at once.

The ffmpeg poster generation (which is a fallback if the browser can't display a poster natively) is also limited to 1 process at a time.

So if multiple videos are uploaded, it might take a tad longer due to this limitation, but at least nothing is crashing.

## Relevant Technical Choices



Updated tests to make them a bit more robust hopefully.

## To-do



## User-facing changes



No visual changes.

## Testing Instructions






This PR can be tested by following these steps:

Video upload:

1. Create a new story.
1. Drag & drop multiple (e.g. 5) MOV files onto the canvas at once
2. Observe the videos uploading
3. The browser should not crash

Video optimization:

1. Go to the media library in WordPress
2. Upload a few MP4 videos
1. Create a new story with a few pages
2. Add all these new MP4 videos to the story
3. Open the pre-publish checklist and open the "Accessibility" tab
4. There should be a note about optimizing videos
5. Optimize all videos
6. The browser should not crash

## Reviews

### Does this PR have a security-related impact?



No

### Does this PR change what data or activity we track or use?



No

### Does this PR have a legal-related impact?



No

## Checklist




---



Fixes #8779
